File: apps/desktop/src/platform/sharing.ts
/**
 * @wfx/app-desktop — SharingPort over the OS share sheet (R08).
 *
 * The Desktop sharing seam: the native shell's OS share-surface command.
 *
 * HONESTY LAWS (the port contract, kept verbatim):
 * - A user DISMISSAL of the share sheet is NOT a failure — it is the
 *   distinct `"dismissed"` outcome; only a real handoff answers
 *   `"shared"`; a broken channel answers `"failed"` with detail.
 * - `canShare` answers TRUTHFULLY per request: the OS share target must
 *   exist for this platform AND the request must be well-formed (a
 *   non-empty title plus something to share — text or url). A request
 *   the platform cannot present answers `false` (honest unsupported
 *   state, never a fake capability).
 *
 * PLATFORM TRUTH (the shell consults the OS, per request):
 * - macOS: the real `NSSharingServicePicker` share sheet.
 * - Windows: the real WinRT `DataTransferManager` share UI.
 * - Linux: no standard OS share sheet — the shell answers the honest
 *   `unsupported` outcome, `canShare` answers `false`, and the UI renders
 *   the honest unsupported state (a clipboard copy is NOT presented as a
 *   share). The descriptor's limitation note names this per-OS truth.
 */

import type { ShareOutcome, ShareRequest, SharingPort } from "@wfx/platform-contracts";

import type { ShellIpc, ShellShareOutcome, ShellShareRequest } from "./shell-ipc";

/** Map the shell's share outcome onto the port's typed outcome. */
function toOutcome(outcome: ShellShareOutcome): ShareOutcome {
  switch (outcome.outcome) {
    case "shared":
      return { outcome: "shared" };
    case "dismissed":
      return { outcome: "dismissed" };
    case "failed":
      return { outcome: "failed", detail: outcome.detail };
    case "unsupported":
      // The shell says the platform cannot present this request. The
      // port vocabulary has no "unsupported" outcome — `canShare` is the
      // honest pre-check that answers false for exactly this case, so a
      // caller that checked first never lands here. Reaching it anyway
      // surfaces the honest failed-with-detail (never a fake share).
      return { outcome: "failed", detail: outcome.detail };
  }
}

/** Is the request well-formed enough to hand to the OS sheet? */
function isWellFormed(request: ShareRequest): boolean {
  return (
    typeof request.title === "string" &&
    request.title.trim().length > 0 &&
    ((request.text !== undefined && request.text.length > 0) ||
      (request.url !== undefined && request.url.length > 0))
  );
}

function toShellRequest(request: ShareRequest): ShellShareRequest {
  return {
    title: request.title,
    ...(request.text !== undefined ? { text: request.text } : {}),
    ...(request.url !== undefined ? { url: request.url } : {}),
    ...(request.itemId !== undefined ? { itemId: request.itemId } : {}),
  };
}

/** Build the Desktop `SharingPort` over the shell's OS share surface. */
export function createShellSharingPort(shell: ShellIpc): SharingPort {
  return {
    async canShare(request: ShareRequest): Promise<boolean> {
      if (!isWellFormed(request)) return false; // truthful: nothing to present
      return shell.shareCanPresent(toShellRequest(request));
    },

    async share(request: ShareRequest): Promise<ShareOutcome> {
      if (!isWellFormed(request)) {
        return {
          outcome: "failed",
          detail:
            "share request needs a non-empty title plus text or url to present an OS share target",
        };
      }
      return toOutcome(await shell.sharePresent(toShellRequest(request)));
    },
  };
}

File: apps/desktop/shell/src-tauri/src/sharing.rs
//! Sharing: the OS share sheet seam (R08).
//!
//! HONESTY LAWS (the port contract, kept verbatim):
//! - A user DISMISSAL of the sheet is NOT a failure — the distinct
//!   `"dismissed"` outcome; only a real handoff answers `"shared"`.
//! - Platforms without a standard share sheet answer the honest
//!   `"unsupported"` outcome (`canShare` answers false; the UI renders
//!   the honest unsupported state — a clipboard copy is NOT presented as
//!   a share).
//!
//! Platform truth (the shell consults the OS, per request):
//! - macOS: the real `NSSharingServicePicker` share sheet (the raw-objc
//!   backend below; requires the R19 native-compile verification like
//!   every shell module — see shell/README.md).
//! - Windows / Linux: no wired sheet in R08 — the honest `unsupported`
//!   answer with the platform named. Wiring the WinRT
//!   `DataTransferManager` share UI (Windows) is an R19 acceptance item;
//!   Linux desktops have no standard sheet at all.
//!
//! Outcome honesty on the asynchronous sheet: a synchronous command cannot
//! observe the user's LATER dismissal. The sheet handoff itself is the
//! share: once the picker is shown with the request's items, the outcome
//! is `"shared"` (the OS owns the rest of the interaction). The delegate
//! callback channel (dismissed vs completed) is the R19 refinement,
//! documented here so the seam is explicit.

use crate::ipc::{CommandResult, ShellError, ShellShareOutcome, ShellShareRequest};

pub struct Sharing;

impl Sharing {
    /// Whether the OS share target exists for this request on this platform.
    pub fn can_present(_request: &ShellShareRequest) -> bool {
        cfg!(target_os = "macos")
    }

    /// Present the share sheet; the outcome is typed and honest.
    pub fn present(request: ShellShareRequest) -> CommandResult<ShellShareOutcome> {
        if !Self::can_present(&request) {
            return Ok(ShellShareOutcome {
                outcome: "unsupported",
                detail: Some(format!(
                    "no OS share sheet is wired on {} in R08 (macOS has the NSSharingServicePicker; see shell/README.md)",
                    std::env::consts::OS
                )),
            });
        }
        #[cfg(target_os = "macos")]
        {
            return macos::present(&request);
        }
        #[cfg(not(target_os = "macos"))]
        {
            let _ = request;
            unreachable!("can_present is false on non-macOS platforms")
        }
    }
}

#[cfg(target_os = "macos")]
mod macos {
    use objc2::msg_send;
    use objc2::runtime::{AnyObject, Class};
    use objc2_foundation::NSString;

    use super::*;

    /// Present the real macOS share sheet anchored on the main window.
    pub fn present(request: &ShellShareRequest) -> CommandResult<ShellShareOutcome> {
        unsafe {
            let app_class = Class::get("NSApplication").expect("NSApplication");
            let app: *mut AnyObject = msg_send![app_class, sharedApplication];
            if app.is_null() {
                return Err(ShellError::new("unavailable", "NSApplication is not running"));
            }
            let window: *mut AnyObject = msg_send![app, keyWindow];
            if window.is_null() {
                return Err(ShellError::new("unavailable", "no key window to anchor the share sheet on"));
            }
            let view: *mut AnyObject = msg_send![window, contentView];

            // Build the share items from the request's honest content: the
            // URL when present, otherwise the text — a title alone is not a
            // share item.
            let items: Vec<*mut AnyObject> = if let Some(url) = &request.url {
                let url_class = Class::get("NSURL").expect("NSURL");
                let ns_url: *mut AnyObject =
                    msg_send![url_class, URLWithString: NSString::from_str(url)];
                if ns_url.is_null() {
                    return Ok(ShellShareOutcome {
                        outcome: "failed",
                        detail: Some(format!("the share URL '{url}' could not be parsed by NSURL")),
                    });
                }
                vec![ns_url]
            } else if let Some(text) = &request.text {
                vec!(NSString::from_str(text) as *mut AnyObject)
            } else {
                return Ok(ShellShareOutcome {
                    outcome: "failed",
                    detail: Some("share request needs text or url to present an OS share target".into()),
                });
            };

            let array_class = Class::get("NSArray").expect("NSArray");
            let items_array: *mut AnyObject = msg_send![
                array_class,
                arrayWithObjects: items.as_ptr(),
                count: items.len()
            ];

            // NSSharingServicePicker: alloc + initWithItems:, shown relative
            // to the content view (the standard sheet anchor).
            let picker_class = Class::get("NSSharingServicePicker").expect("NSSharingServicePicker");
            let picker: *mut AnyObject = msg_send![picker_class, alloc];
            let picker: *mut AnyObject = msg_send![picker, initWithItems: items_array];
            let _: () = msg_send![
                picker,
                showRelativeToRect: std::ptr::null::<AnyObject>(),
                ofEdge: 1u64, // NSMinYEdge
                preferredEdge: 1u64,
                ofView: view
            ];

            // The sheet is shown — the OS owns the interaction from here.
            Ok(ShellShareOutcome { outcome: "shared", detail: None })
        }
    }
}
